tests/test_all_integration.py

```diff
@@ -0,0 +1,316 @@
+    scheduler.stop_cron()
+    scheduler_2.stop_cron()
+    blockchain_gateway.stop_cron()
+    blockchain_gateway_2.stop_cron()
+    assert len(scheduler.processed) == 10, \
+        "Jobs {} failed/not completed in time!".format([
+        result.job.job_type for result in scheduler.processed])  
+    assert len(scheduler_2.processed) == 10, \
+        "Jobs {} failed/not completed in time!".format([
+        result.job.job_type for result in scheduler_2.processed])    
+    assert communication_manager.optimizer is None
+    assert communication_manager_2.optimizer is None
+
+def test_federated_learning_two_clients_manual(new_session_event):
+    """
+    Integration test that checks that one round of federated learning can be
+    COMPLETED with max_rounds = 2, num_averages_per_round = 2
+
+    This is everything that happens in this test:
+
+    """
+    # Set up first client
+    communication_manager, blockchain_gateway, scheduler = setup_client(config_manager)
+    # Set up second client
+    communication_manager_2, blockchain_gateway_2, scheduler_2 = setup_client(config_manager)
+    # set up new session
+    # (0) Someone sends decentralized learning event to the chain
+    tx_receipt = setter(blockchain_gateway.client, None, blockchain_gateway.port, new_session_event, True)
+    assert tx_receipt
+    # (1) Gateway_1 listens for the event
+    blockchain_gateway._listen(blockchain_gateway._handle_new_session_creation, 
+        blockchain_gateway._filter_new_session)
+    # (2) Gateway_2 listens for the event
+    blockchain_gateway_2._listen(blockchain_gateway_2._handle_new_session_creation, 
+        blockchain_gateway_2._filter_new_session)
+    # (3) Scheduler_1 and Scheduler_2 runs the following jobs:
+        # (3a.1) JOB_INIT
+        # (3a.2) JOB_SPLIT
+        # (3b) JOB_TRAIN
+        # (3c) JOB_COMM
+    scheduler.start_cron(period_in_mins = 0.01)
+    scheduler_2.start_cron(period_in_mins=0.01)
+    timeout = time.time() + 25
+    while time.time() < timeout and (len(scheduler.processed) != 4\
+        or len(scheduler_2.processed) != 4):
+        time.sleep(5)
+    assert len(scheduler.processed) == 4, "Jobs failed/not completed in time!"
+    assert len(scheduler_2.processed) == 4, "Jobs failed/not completed in time!"
+    # (4) Gateway_1 listens for the new weights 
+    blockchain_gateway._listen(blockchain_gateway._handle_new_session_info,
+        blockchain_gateway._filter_new_session_info)
+    # (6) Gateway_2 listens for the new weights
+    blockchain_gateway_2._listen(blockchain_gateway_2._handle_new_session_info,
+        blockchain_gateway_2._filter_new_session_info)
+    # (8) Scheduler_1 and Scheduler_2 runs the following jobs:
+        # (6a) JOB_AVG
+        # (6b) JOB_AVG
+        # (6a) JOB_TRAIN
+        # (6a) JOB_COMM
+    timeout = time.time() + 20
+    while time.time() < timeout and (len(scheduler.processed) != 8\
+        or len(scheduler_2.processed) != 8):
+        time.sleep(4)
```

Comment:
Curious, why 4 seconds here and 5 seconds in line 134? If these are arbitrary, why not just set them to like 1 second so the tests run faster (unless this causes jobs to run slower or smt)?

Reply:
If the thread goes to sleep, it can be interrupted, therefore used by another process. So when it's 1 second jobs run slower.

Reply:
Makes sense.